### ISSUE TITLE
prepSim Error: 2/0 samples retained and length of 'dimnames' not equal to array extent

### ISSUE THREAD

**Issue by domi84** [HelenaLC/muscat#59]
Congrats for the publication! 
I installed muscat and I was trying to simulate data, from here:
https://bioconductor.org/packages/release/bioc/vignettes/muscat/inst/doc/simulation.html#prepsim-preparing-data-for-simulation

I tried first with the tutorial dataset, and it works
```
> data(sce)
> ref <- prepSim(sce)
Argument `group_keep` unspecified; defaulting to retaining “ctrl”-group samples.
Filtering...
- 1157/1267 genes and 750/1556 cells retained.
- 2/5 subpopulations and 2/2 samples retained.
Estimating gene and cell parameters...
```

Then I tried with a dataset I have, but I got an error:
```
#Convert it from Seurat
>DefaultAssay(DC) <- "RNA"
>DC.sce <- as.SingleCellExperiment(DC)

>DC.sce$sample_id <- DC.sce$sample
>DC.sce$group_id <- DC.sce$group
>DC.sce$cluster_id <- DC.sce$DCType

> ref <- prepSim(DC.sce, group_keep="Active D")
Filtering...
- 7054/27248 genes and 1380/2236 cells retained.
- 1/6 subpopulations and 2/0 samples retained.
Estimating gene and cell parameters...
Error in dimnames(x) <- dn :
  length of 'dimnames' [2] not equal to array extent
> ref <- prepSim(DC.sce, group_keep="Healthy")
Filtering...
- 4610/27248 genes and 319/2236 cells retained.
- 1/6 subpopulations and 2/0 samples retained.
Estimating gene and cell parameters...
Error in dimnames(x) <- dn :
  length of 'dimnames' [2] not equal to array extent
```

Any idea what I am doing wrong?
Thanks

```
> sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /export/home2/ds286q/.conda/envs/muscat_stable/lib/libopenblasp-r0.3.12.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets
[8] methods   base

other attached packages:
 [1] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
 [3] Biobase_2.50.0              GenomicRanges_1.42.0
 [5] GenomeInfoDb_1.26.4         IRanges_2.24.1
 [7] S4Vectors_0.28.1            BiocGenerics_0.36.0
 [9] MatrixGenerics_1.2.1        matrixStats_0.58.0
[11] dplyr_1.0.5                 muscat_1.5.0
[13] SeuratObject_4.0.0          Seurat_4.0.0

loaded via a namespace (and not attached):
  [1] utf8_1.1.4                reticulate_1.18
  [3] tidyselect_1.1.0          lme4_1.1-26
  [5] RSQLite_2.2.3             AnnotationDbi_1.52.0
  [7] htmlwidgets_1.5.3         grid_4.0.3
  [9] BiocParallel_1.24.1       Rtsne_0.15
 [11] munsell_0.5.0             codetools_0.2-18
 [13] ica_1.0-2                 statmod_1.4.35
 [15] future_1.21.0             miniUI_0.1.1.1
 [17] colorspace_2.0-0          rstudioapi_0.13
 [19] ROCR_1.0-11               tensor_1.5
 [21] listenv_0.8.0             GenomeInfoDbData_1.2.4
 [23] polyclip_1.10-0           bit64_4.0.5
 [25] glmmTMB_1.0.2.1           parallelly_1.23.0
 [27] vctrs_0.3.6               generics_0.1.0
 [29] R6_2.5.0                  doParallel_1.0.16
 [31] ggbeeswarm_0.6.0          clue_0.3-58
 [33] rsvd_1.0.3                locfit_1.5-9.4
 [35] bitops_1.0-6              spatstat.utils_2.0-0
 [37] cachem_1.0.4              DelayedArray_0.16.2
 [39] assertthat_0.2.1          promises_1.2.0.1
 [41] scales_1.1.1              beeswarm_0.3.1
 [43] gtable_0.3.0              beachmat_2.6.4
 [45] Cairo_1.5-12.2            globals_0.14.0
 [47] goftest_1.2-2             rlang_0.4.10
 [49] genefilter_1.72.1         GlobalOptions_0.1.2
 [51] splines_4.0.3             TMB_1.7.19
 [53] lazyeval_0.2.2            broom_0.7.5
 [55] reshape2_1.4.4            abind_1.4-5
 [57] backports_1.2.1           httpuv_1.5.5
 [59] tools_4.0.3               ggplot2_3.3.3
 [61] gplots_3.1.1              ellipsis_0.3.1
 [63] RColorBrewer_1.1-2        ggridges_0.5.3
 [65] Rcpp_1.0.6                plyr_1.8.6
 [67] sparseMatrixStats_1.2.1   progress_1.2.2
 [69] zlibbioc_1.36.0           purrr_0.3.4
 [71] RCurl_1.98-1.2            prettyunits_1.1.1
 [73] rpart_4.1-15              deldir_0.2-10
 [75] pbapply_1.4-3             GetoptLong_1.0.5
 [77] viridis_0.5.1             cowplot_1.1.1
 [79] zoo_1.8-9                 ggrepel_0.9.1
 [81] cluster_2.1.1             colorRamps_2.3
 [83] variancePartition_1.20.0  magrittr_2.0.1
 [85] data.table_1.14.0         scattermore_0.7
 [87] lmerTest_3.1-3            circlize_0.4.12
 [89] lmtest_0.9-38             RANN_2.6.1
 [91] fitdistrplus_1.1-3        hms_1.0.0
 [93] patchwork_1.1.1           mime_0.10
 [95] xtable_1.8-4              pbkrtest_0.5.1
 [97] XML_3.99-0.5              gridExtra_2.3
 [99] shape_1.4.5               compiler_4.0.3
[101] scater_1.18.6             tibble_3.1.0
[103] KernSmooth_2.23-18        crayon_1.4.1
[105] minqa_1.2.4               htmltools_0.5.1.1
[107] mgcv_1.8-34               later_1.1.0.1
[109] tidyr_1.1.3               geneplotter_1.68.0
[111] DBI_1.1.1                 ComplexHeatmap_2.6.2
[113] MASS_7.3-53.1             boot_1.3-27
[115] Matrix_1.3-2              igraph_1.2.6
[117] pkgconfig_2.0.3           numDeriv_2016.8-1.1
[119] plotly_4.9.3              scuttle_1.0.4
[121] foreach_1.5.1             annotate_1.68.0
[123] vipor_0.4.5               blme_1.0-5
[125] XVector_0.30.0            stringr_1.4.0
[127] digest_0.6.27             sctransform_0.3.2
[129] RcppAnnoy_0.0.18          spatstat.data_2.0-0
[131] leiden_0.3.7              uwot_0.1.10
[133] edgeR_3.32.1              DelayedMatrixStats_1.12.3
[135] gtools_3.8.2              shiny_1.6.0
[137] rjson_0.2.20              nloptr_1.2.2.2
[139] lifecycle_1.0.0           nlme_3.1-152
[141] jsonlite_1.7.2            BiocNeighbors_1.8.2
[143] viridisLite_0.3.0         limma_3.46.0
[145] fansi_0.4.2               pillar_1.5.1
[147] lattice_0.20-41           fastmap_1.1.0
[149] httr_1.4.2                survival_3.2-7
[151] glue_1.4.2                spatstat_1.64-1
[153] iterators_1.0.13          png_0.1-7
[155] bit_4.0.4                 stringi_1.5.3
[157] blob_1.2.1                DESeq2_1.30.1
[159] BiocSingular_1.6.0        caTools_1.18.1
[161] memoise_2.0.0             irlba_2.3.3
[163] future.apply_1.7.0
```

**Comment by HelenaLC** [comment]
Could you print `DC.sc` and `head(colData(DC.sc))`? Thanks!

**Comment by domi84** [comment]
here, thanks:
```
> DC.sce
class: SingleCellExperiment
dim: 27248 2236
metadata(0):
assays(2): counts logcounts
rownames(27248): AL627309.1 AL669831.5 ... DUOXA2 RP11-58A18.2
rowData names(0):
colnames(2236): ACAGCCGGTATGGAAT_1_1 ACGGTCGGTATTGACC_1_1 ...
  SA225_TGACTTTTCACAATGC_2 SA225_TGACTTTTCCTTGCCA_2
colData names(23): orig.ident nCount_RNA ... group_id cluster_id
reducedDimNames(3): PCA UMAP PHATE
altExpNames(0):

> head(colData(DC.sce))
DataFrame with 6 rows and 23 columns
                      orig.ident nCount_RNA nFeature_RNA           origin
                     <character>  <numeric>    <integer>      <character>
ACAGCCGGTATGGAAT_1_1 SA271bBlood       4971         1733 Peripheral Blood
ACGGTCGGTATTGACC_1_1 SA271bBlood       7520         2236 Peripheral Blood
ACTCTCGTCTATCGCC_1_1 SA271bBlood       2941         1288 Peripheral Blood
ACTGTCCAGCACCGAA_1_1 SA271bBlood       4096         1650 Peripheral Blood
ACTGTGAAGAGATGCC_1_1 SA271bBlood       2323         1033 Peripheral Blood
ACTTTGTTCACACCCT_1_1 SA271bBlood       3705         1637 Peripheral Blood
                          sample     group percent.mt    location       state
                     <character> <ordered>  <numeric> <character> <character>
ACAGCCGGTATGGAAT_1_1       SA271 Active D    9.21344          NA          NA
ACGGTCGGTATTGACC_1_1       SA271 Active D    5.19947          NA          NA
ACTCTCGTCTATCGCC_1_1       SA271 Active D   17.57905          NA          NA
ACTGTCCAGCACCGAA_1_1       SA271 Active D    4.44336          NA          NA
ACTGTGAAGAGATGCC_1_1       SA271 Active D    6.88765          NA          NA
ACTTTGTTCACACCCT_1_1       SA271 Active D    7.82726          NA          NA
                     completeIntegration_snn_res.0.2 seurat_clusters
                                            <factor>        <factor>
ACAGCCGGTATGGAAT_1_1                               3               2
ACGGTCGGTATTGACC_1_1                               0               5
ACTCTCGTCTATCGCC_1_1                               3               2
ACTGTCCAGCACCGAA_1_1                               3               2
ACTGTGAAGAGATGCC_1_1                               3               2
ACTTTGTTCACACCCT_1_1                               3               2
                     completeIntegration_snn_res.1.2  grouporigin
                                            <factor>    <ordered>
ACAGCCGGTATGGAAT_1_1                              1  Active D PB
ACGGTCGGTATTGACC_1_1                              16 Active D PB
ACTCTCGTCTATCGCC_1_1                              1  Active D PB
ACTGTCCAGCACCGAA_1_1                              1  Active D PB
ACTGTGAAGAGATGCC_1_1                              19 Active D PB
ACTTTGTTCACACCCT_1_1                              1  Active D PB
                     nCount_integratedByGroup nFeature_integratedByGroup
                                    <numeric>                  <integer>
ACAGCCGGTATGGAAT_1_1                        0                          0
ACGGTCGGTATTGACC_1_1                        0                          0
ACTCTCGTCTATCGCC_1_1                        0                          0
ACTGTCCAGCACCGAA_1_1                        0                          0
ACTGTGAAGAGATGCC_1_1                        0                          0
ACTTTGTTCACACCCT_1_1                        0                          0
                     completeIntegration_snn_res.0.6
                                            <factor>
ACAGCCGGTATGGAAT_1_1                               2
ACGGTCGGTATTGACC_1_1                               5
ACTCTCGTCTATCGCC_1_1                               2
ACTGTCCAGCACCGAA_1_1                               2
ACTGTGAAGAGATGCC_1_1                               2
ACTTTGTTCACACCCT_1_1                               2
                     completeIntegration_snn_res.0.4 CellTypes      DCType
                                            <factor> <ordered> <character>
ACAGCCGGTATGGAAT_1_1                               2  B1           B
ACGGTCGGTATTGACC_1_1                               0  A0           A
ACTCTCGTCTATCGCC_1_1                               2  B1           B
ACTGTCCAGCACCGAA_1_1                               2  B1           B
ACTGTGAAGAGATGCC_1_1                               2  B1           B
ACTTTGTTCACACCCT_1_1                               2  B1           B
                        ident   sample_id  group_id cluster_id
                     <factor> <character> <ordered>  <ordered>
ACAGCCGGTATGGAAT_1_1 B1       SA271 Active D   B1
ACGGTCGGTATTGACC_1_1 A0       SA271 Active D   A0
ACTCTCGTCTATCGCC_1_1 B1       SA271 Active D   B1
ACTGTCCAGCACCGAA_1_1 B1       SA271 Active D   B1
ACTGTGAAGAGATGCC_1_1 B1       SA271 Active D   B1
ACTTTGTTCACACCCT_1_1 B1       SA271 Active D   B1
>
```

**Comment by HelenaLC** [comment]
Hm. I assume this is indeed a bug as I just tried to reproduce this strange message `0/2 samples retained` - and succeeded. Could you try setting the following and seeing if the error disappears? I fear there's a check missing...
```
for (. in c("cluster_id", "sample_id", "group_id"))
  sce[[.]] <- factor(sce[[.]])
```

**Comment by domi84** [comment]
Thanks, it worked!

```
> for (. in c("cluster_id", "sample_id", "group_id"))
+   DC.sce[[.]] <- factor(DC.sce[[.]])
> ref <- prepSim(DC.sce)
Argument `group_keep` unspecified; defaulting to retaining “Healthy”-group samples.
Filtering...
- 4610/27248 genes and 319/2236 cells retained.
- 1/6 subpopulations and 2/5 samples retained.
Estimating gene and cell parameters...
>
```

**Comment by HelenaLC** [comment]
Fixed in version 1.5.1 available through GitHub master branch now, and Bioc devel in a couple days.
